lib/charms/opensearch/v0/opensearch_users.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""OpenSearch user helper functions.
+
+These functions wrap around some API calls used for user management.
+"""
+
+import logging
+from typing import Any, Dict, List
+
+from charms.opensearch.v0.opensearch_distro import OpenSearchDistribution
+
+logger = logging.getLogger(__name__)
+
+
+USER_ENDPOINT = "/_plugins/_security/api/internalusers"
+ROLE_ENDPOINT = "/_plugins/_security/api/roles"
+
+
+class OpenSearchUserMgmtError(Exception):
+    """Base exception class for OpenSearch user management errors."""
+
+
+def create_role(
+    opensearch: OpenSearchDistribution,
+    role_name: str,
+    permissions: Dict = {},
+    action_groups: Dict = {},
+) -> Dict[str, Any]:
+    """Creates a role with the given permissions.
+
+    This method assumes the dicts provided are valid opensearch config. If not, raises
+    OpenSearchUserMgmtError.
+
+    Args:
+        opensearch: Opensearch distribution object, used to send requests
+        role_name: name of the role
+        permissions: A valid dict of existing opensearch permissions.
+        action_groups: A valid dict of existing opensearch action groups.
+
+    Raises:
+        OpenSearchUserMgmtError: If the role creation request fails.
+
+    Returns:
+        Output of the role creation request.
+    """
+    put_role_resp = opensearch.request(
+        "PUT",
+        f"{ROLE_ENDPOINT}/{role_name}",
+        {**permissions, **action_groups},
+    )
+    # enable this role
+    logger.debug(put_role_resp)
+    if put_role_resp.get("status") != "OK":
+        raise OpenSearchUserMgmtError()
+    return put_role_resp
+
+
+def remove_role(opensearch: OpenSearchDistribution, role_name: str) -> Dict[str, Any]:
+    """Remove the given role from opensearch distribution.
+
+    Args:
+        opensearch: Opensearch distribution object, used to send requests
+        role_name: name of the role to be removed.
+
+    Raises:
+        OpenSearchUserMgmtError: If the request fails.
+
+    Returns:
+        Output of the request.
+    """
+    resp = opensearch.request("DELETE", f"{ROLE_ENDPOINT}/{role_name}")
+    logger.debug(resp)
+    # check if I have to disable roles before removal
+    if resp.get("status") != "OK":
+        raise OpenSearchUserMgmtError()
+    return resp
+
+
+def create_user(
+    opensearch: OpenSearchDistribution,
+    username: str,
```

Comment:
nit: `user_name` so that we stay consistent with the 2 words variables / params (i.e: `hashed_pwd`)